libc-bottom-half/sources/chdir.c

```diff
@@ -62,15 +63,41 @@ int chdir_legacy(const char *path)
     __wasilibc_cwd_lock();
     char *prev_cwd = __wasilibc_cwd;
     __wasilibc_cwd = new_cwd;
-    __wasilibc_cwd_unlock();
 
     if (__wasilibc_cwd_mallocd)
         free(prev_cwd);
+
     __wasilibc_cwd_mallocd = 1;
+    __wasilibc_cwd_unlock();
     return 0;
 }
 
 static const char *make_absolute(const char *path) {
+    // if the libc has not synced with the runtime yet, call into the runtime to get the cwd
+    if (!__wasilibc_cwd_is_synced) {
+        char cwd[256];
+        if (getcwd(cwd, sizeof(cwd)) != NULL) {
+            char *new_cwd = malloc(strlen(cwd) + 1);
+            if (new_cwd != NULL) {
+                strcpy(new_cwd, cwd);
+            }
+
+            __wasilibc_cwd_lock();
+            char *prev_cwd = __wasilibc_cwd;
+            __wasilibc_cwd = new_cwd;
+
+            if (__wasilibc_cwd_mallocd)
+                free(prev_cwd);
+
+            __wasilibc_cwd_is_synced = 1; // we synced the cwd
```

Comment:
Doesn't this block technically only run once, since we're setting `is_synced` here? There'd be no next time to take into account.